nixos/modules/services/cluster/kubernetes.nix

```diff
@@ -527,109 +640,678 @@ in {
+            enable = cfg.etcd.keyFile != null;
+            hostPath.path = cfg.etcd.keyFile;
+            name = "etcd-key";
+          }];
+        };
+      };
+
+      services.kubernetes.kubelet.networkPlugin = "cni";
+      services.kubernetes.kubelet.cniPluginPackages = [pkgs.calico-cni];
+      services.calico.enable = mkDefault true;
+
+      environment.etc."cni/net.d/10-calico.conf".text = builtins.toJSON {
+        name = "calico-k8s-network";
+        type = "calico";
+        etcd_endpoints = concatStringsSep "," cfg.etcd.servers;
+        etcd_key_file = cfg.etcd.keyFile;
+        etcd_cert_file = cfg.etcd.certFile;
+        etcd_ca_cert_file = cfg.etcd.caFile;
+        log_level = "info";
+        policy = {
+          type = "k8s";
+        };
+        kubernetes = {
+          k8s_api_root = "http://${cfg.master}";
+          kubeconfig = kubeconfig;
+        };
+        ipam = {
+          type = "host-local";
+          subnet = "usePodCidr";
         };
       };
     })
 
     (mkIf (any (el: el == "master") cfg.roles) {
+      virtualisation.docker.enable = mkDefault true;
+      services.kubernetes.kubelet.enable = mkDefault true;
+      services.kubernetes.kubelet.allowPrivileged = mkDefault true;
+      services.kubernetes.kubelet.registerSchedulable = mkOverride 1000 false;
       services.kubernetes.apiserver.enable = mkDefault true;
       services.kubernetes.scheduler.enable = mkDefault true;
       services.kubernetes.controllerManager.enable = mkDefault true;
-      services.kubernetes.kube2sky.enable = mkDefault true;
+      services.etcd.enable = mkDefault (cfg.etcd.servers == ["http://127.0.0.1:2379"]);
     })
 
     (mkIf (any (el: el == "node") cfg.roles) {
       virtualisation.docker.enable = mkDefault true;
       services.kubernetes.kubelet.enable = mkDefault true;
       services.kubernetes.proxy.enable = mkDefault true;
-    })
-
-    (mkIf (any (el: el == "node" || el == "master") cfg.roles) {
-      services.etcd.enable = mkDefault true;
-
-      services.skydns.enable = mkDefault true;
-      services.skydns.domain = mkDefault cfg.kubelet.clusterDomain;
+      services.kubernetes.dns.enable = mkDefault true;
     })
 
     (mkIf (
         cfg.apiserver.enable ||
         cfg.scheduler.enable ||
         cfg.controllerManager.enable ||
         cfg.kubelet.enable ||
-        cfg.proxy.enable
+        cfg.proxy.enable ||
+        cfg.dns.enable
     ) {
-      systemd.services.kubernetes-setup = {
-        description = "Kubernetes setup.";
-        serviceConfig.Type = "oneshot";
-        script = ''
-          mkdir -p /var/run/kubernetes
-          chown kubernetes /var/lib/kubernetes
-
-          rm ${cfg.dataDir}/.dockercfg || true
-          ln -fs ${pkgs.writeText "kubernetes-dockercfg" cfg.dockerCfg} ${cfg.dataDir}/.dockercfg
-        '';
-      };
-
-      services.kubernetes.package = mkDefault pkgs.kubernetes;
+      systemd.tmpfiles.rules = [
+        "d /run/kubernetes 0755 root root -"
```

Comment:
They point to same location(symlink) on nixos, but will change as it is inconsistent.